File: Template_Cpp/1_Basic/MyCar.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>
#include "RuleBasedDriving.h"

using namespace std;
using namespace Car;

bool is_debug = false;
bool is_accident = false;
int accident_count = 0;
int recovery_count = 0;

bool enable_api_control = true;	// true(Controlled by code) /false(Controlled by keyboard)

ControlValues control_driving(CarStateValues sensing_info)
{
	if (is_debug)
	{
		cout << "=========================================================" << endl;
		cout << "[MyCar] to middle: " << sensing_info.to_middle << endl;

		cout << "[MyCar] collided: " << sensing_info.collided << endl;
		cout << "[MyCar] car speed: " << sensing_info.speed << "km/h" << endl;

		cout << "[MyCar] is moving forward: " << sensing_info.moving_forward << endl;
		cout << "[MyCar] moving angle: " << sensing_info.moving_angle << endl;
		cout << "[MyCar] lap_progress: " << sensing_info.lap_progress << endl;

		cout << "[MyCar] track_forward_angles: ";
		for (int i = 0; i < sensing_info.track_forward_angles.size(); i++)
		{
			cout << sensing_info.track_forward_angles[i] << ", ";
		}
		cout << endl;

		cout << "[MyCar] track_forward_obstacles: ";
		for (int i = 0; i < sensing_info.track_forward_obstacles.size(); i++)
		{
			cout << "{dist: " << sensing_info.track_forward_obstacles[i].dist 
				 << ", to_middle: " << sensing_info.track_forward_obstacles[i].to_middle << "}, ";
		}
		cout << endl;

		cout << "[MyCar] opponent_cars_info: ";
		for (int i = 0; i < sensing_info.opponent_cars_info.size(); i++)
		{
			cout << "{dist: " << sensing_info.opponent_cars_info[i].dist 
				 << ", to_middle: " << sensing_info.opponent_cars_info[i].to_middle 
				 << ", speed: " << sensing_info.opponent_cars_info[i].speed << "km/h}, ";
		}
		cout << endl;

		cout << "[MyCar] distance_to_way_points: ";
		for (int i = 0; i < sensing_info.distance_to_way_points.size(); i++)
		{
			cout << sensing_info.distance_to_way_points[i] << ", ";
		}
		cout << endl;

		cout << "=========================================================" << endl;
	}

	ControlValues car_controls;
	// ===========================================================
	// Area for writing code about driving rule ==================
	// ===========================================================
	// Editing area starts from here
	//

	// 도로의 실제 폭의 1/2 로 계산됨
	float half_load_width = sensing_info.half_road_limit - 1.25f;

	// 차량 핸들 조정을 위해 참고할 전방의 커브 값 가져오기
	int angle_num = (int)(sensing_info.speed / 45);
	float ref_angle = angle_num > 0 ? sensing_info.track_forward_angles[angle_num] : 0;

	// 차량의 차선 중앙 정렬을 위한 미세 조정 값 계산
	float middle_add = (sensing_info.to_middle / 80) * -1;

	// 전방의 커브 각도에 따라 throttle 값을 조절하여 속도를 제어함
	float throttle_factor = 0.6f / (abs(ref_angle) + 0.1f);
	if (throttle_factor > 0.11f) throttle_factor = 0.11f;   // throttle 값을 최대 0.81 로 설정
	float set_throttle = 0.7f + throttle_factor;
	if (sensing_info.speed < 60) set_throttle = 0.9f;  // 속도가 60Km/h 이하인 경우 0.9 로 설정
	if (sensing_info.speed > 80) set_throttle = 0.6f;  // 최대속도를 80km/h로 설정
	
	// 차량의 Speed 에 따라서 핸들을 돌리는 값을 조정함
	float steer_factor = sensing_info.speed * 1.5f;
	if (sensing_info.speed > 70) steer_factor = sensing_info.speed * 0.85f;
	if (sensing_info.speed > 100) steer_factor = sensing_info.speed * 0.7f;

	// (참고할 전방의 커브 - 내 차량의 주행 각도) / (계산된 steer factor) 값으로 steering 값을 계산
	float set_steering = (ref_angle - sensing_info.moving_angle) / (steer_factor + 0.001f);

	// 차선 중앙정렬 값을 추가로 고려함
	set_steering += middle_add;

	//// 긴급 및 예외 상황 처리 ////////////////////////////////////////////////////////////////////////////////////////
	bool full_throttle = true;
	bool emergency_brake = false;

	// 전방 커브의 각도가 큰 경우 속도를 제어함
	// 차량 핸들 조정을 위해 참고하는 커브 보다 조금 더 멀리 참고하여 미리 속도를 줄임
	int road_range = (int)(sensing_info.speed / 30);
	for (int i = 0; i < road_range; i++) {
		float fwd_angle = abs(sensing_info.track_forward_angles[i]);
		if (fwd_angle > 45) {   // 커브가 45도 이상인 경우 brake, throttle 을 제어
			full_throttle = false;
		}
		if (fwd_angle > 80) {   // 커브가 80도 이상인 경우 steering 까지 추가로 제어
			emergency_brake = true;
			break;
		}
	}
	
	// brake, throttle 제어
	float set_brake = 0.0f;
	if (!full_throttle) {
		if (sensing_info.speed > 100) {
			set_brake = 0.3f;
		}
		if (sensing_info.speed > 120) {
			set_throttle = 0.7f;
			set_brake = 0.7f;
		}
		if (sensing_info.speed > 130) {
			set_throttle = 0.5f;
			set_brake = 1.0f;
		}
	}

	// steering 까지 추가로 제어
	if (emergency_brake) {
		if (set_steering > 0) {
			set_steering += 0.3f;
		}
		else {
			set_steering -= 0.3f;
		}
	}


	////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	// Moving straight forward
	car_controls.steering = set_steering;
	car_controls.throttle = set_throttle;
	car_controls.brake = set_brake;

	if (is_debug)
	{
		cout << "[MyCar] steering:" << car_controls.steering << ", throttle:" << car_controls.throttle
			 << ", brake:" << car_controls.brake << endl;
	}
	//
	// Editing area ends
	// ===========================================================

	return car_controls;
}

int main()
{
	// ===========================================================
	// Don't remove below area. ==================================
	// ===========================================================
	cout << "[MyCar] Start Bot! (CPP)" << endl;

	int return_code = StartDriving(control_driving, enable_api_control);

	cout << "[MyCar] End Bot! (CPP)" << endl;

	// ==========================================================

	return return_code;
}

File: hojong/Bot_Python/my_car.py
from DrivingInterface.drive_controller import DrivingController
import math

class DrivingClient(DrivingController):
    def __init__(self):
        # =========================================================== #
        #  Area for member variables =============================== #
        # =========================================================== #
        # Editing area starts from here
        #

        self.is_debug = False

        # api or keyboard
        self.enable_api_control = True # True(Controlled by code) /False(Controlled by keyboard)
        super().set_enable_api_control(self.enable_api_control)

        #
        # Editing area ends
        # ==========================================================#
        super().__init__()

    def merge_direction(self, sensing_info):
        direction = [0,0,0,0,0] #forward len, angle_sum, len, angle_sum2, len2
        second = False
        angle_sum = 0

        
        for i in range(0,len(sensing_info.track_forward_angles)):
            angle = sensing_info.track_forward_angles[i]
            cur_dir = angle
            if cur_dir != 0:
                cur_dir /= abs(cur_dir)
            idx = 1
            
            angle_sum += angle
            if second:
                idx += 2
            if angle_sum < 20 and angle > -20 and direction[2] == 0:
                direction[0] += 1
            else:
                if angle * direction[idx] < 0 or direction[idx+1] >= 5 or direction[idx] > 100:
                    if second:
                        break
                    second = True
                if direction[1] == 0:
                    direction[1] = angle_sum
                    direction[1] += direction[0]
                    for a in range(0,i):
                        if sensing_info.track_forward_angles[i-a] * angle_sum > 0 and abs(sensing_info.track_forward_angles[i-a]) > 5:
                            direction[0] -= 1
                        else:
                            break
                direction[idx] += angle
                direction[idx+1] += 1
                
        return direction
    
    def control_driving(self, car_controls, sensing_info):

        # =========================================================== #
        # Area for writing code about driving rule ================= #
        # =========================================================== #
        # Editing area starts from here
        #
        #self.is_debug = True
        if self.is_debug:
            print("=========================================================")
            print("[MyCar] to middle: {}".format(sensing_info.to_middle))

            print("[MyCar] collided: {}".format(sensing_info.collided))
            print("[MyCar] car speed: {} km/h".format(sensing_info.speed))

            print("[MyCar] is moving forward: {}".format(sensing_info.moving_forward))
            print("[MyCar] moving angle: {}".format(sensing_info.moving_angle))
            print("[MyCar] lap_progress: {}".format(sensing_info.lap_progress))

            print("[MyCar] track_forward_angles: {}".format(sensing_info.track_forward_angles))
            print("[MyCar] track_forward_obstacles: {}".format(sensing_info.track_forward_obstacles))
            print("[MyCar] opponent_cars_info: {}".format(sensing_info.opponent_cars_info))
            print("[MyCar] distance_to_way_points: {}".format(sensing_info.distance_to_way_points))
            print("=========================================================")

        ###########################################################################

        # Moving straight forward

        steer_sum = 0
        dist_sum = 0
        
        #for i in range(0,20):
        #    steer_sum += (200-sensing_info.distance_to_way_points[i]) * sensing_info.track_forward_angles[i]
        #    dist_sum += 200-sensing_info.distance_to_way_points[i]

        #print(steer_sum/dist_sum)

        direction_info = self.merge_direction(sensing_info)
        #car_controls.steering = -sensing_info.moving_angle/90.0 + steer_avg + forward_five
        #print(car_controls.steering)
        car_controls.steering = 0
        #print(direction_info)
        car_controls.steering -= sensing_info.moving_angle/90
        
        curve_start = (sensing_info.speed-80) // 10

        if curve_start > 5:
            curve_start = 5
        if curve_start < 1:
            curve_start = 1
        
        #print(curve_start, direction_info)
        if direction_info[0] > curve_start:
            #print('hihi')
            #print(-sensing_info.to_middle -direction_info[1][0]*4)
            if direction_info[2] > 0:
                #avg_curve = abs((direction_info[1]+direction_info[3])/(90*(direction_info[2]+direction_info[4])))
                avg_curve = abs(direction_info[1]/(60*direction_info[2]))
                
                if avg_curve > 4:
                    avg_curve = 4
                #modifier = 1+avg_curve
                modifier = 4+avg_curve
                if direction_info[1] < 0:
                    modifier *= -1
                car_controls.steering += math.atan((-sensing_info.to_middle-modifier)/((direction_info[0])*10))*180/math.pi/90
            else:
                car_controls.steering += math.atan(-sensing_info.to_middle/(direction_info[0]*10))*180/math.pi/90
        #elif direction_info[0] > curve_start-2:
        #    car_controls.steering += math.atan(-sensing_info.to_middle/(direction_info[0]*10))*180/math.pi/150
        else:
            #print(direction_info)
            next_forward = sum(sensing_info.track_forward_angles[:direction_info[0]+1])
            car_controls.steering += (direction_info[1]/90 + next_forward/90)/(direction_info[2]+direction_info[0]+1)
            #car_controls.steering += (direction_info[1]/90)/(direction_info[2])
            #car_controls.steering += next_four/90
            #car_controls.steering += direction_info[1]/(90*direction_info[2])
            if direction_info[4] > 0:
                avg_curve = abs(direction_info[3]/(60*direction_info[4]))
                
                if avg_curve > 4:
                    avg_curve = 4
                #modifier = 1+avg_curve
                modifier = 4+avg_curve
                if direction_info[3] < 0:
                    modifier *= -1
                #print(math.atan((-sensing_info.to_middle/2 - modifier)/(direction_info[2]*10))*180/math.pi/90)
                car_controls.steering += math.atan((-sensing_info.to_middle - modifier)/(direction_info[2]*10))*180/math.pi/90
            else:
                car_controls.steering += math.atan(-sensing_info.to_middle/(direction_info[2]*10))*180/math.pi/90

        #next_angle = sensing_info.moving_angle+car_controls.steering*90
        #next_to_middle = math.sin(next_angle*math.pi/180)*sensing_info.speed/10
        #print(next_to_middle)
        #car_controls.steering += math.atan(-next_to_middle/100)*180/math.pi/90

        car_controls.throttle = 1
        car_controls.brake = 0
        if abs(car_controls.steering)*sensing_info.speed > 60:
            car_controls.throttle = 0
            car_controls.brake = abs(car_controls.steering)/1.5
            
        #print('steer', car_controls.steering, 'break', car_controls.brake)
            
        if self.is_debug:
            print("[MyCar] steering:{}, throttle:{}, brake:{}"\
                  .format(car_controls.steering, car_controls.throttle, car_controls.brake))

        #
        # Editing area ends
        # ==========================================================#
        return car_controls


    # ============================
    # If you have NOT changed the <settings.json> file
    # ===> player_name = ""
    #
    # If you changed the <settings.json> file
    # ===> player_name = "My car name" (specified in the json file)  ex) Car1
    # ============================
    def set_player_name(self):
        player_name = ""
        return player_name


if __name__ == '__main__':
    print("[MyCar] Start Bot! (PYTHON)")

    client = DrivingClient()
    return_code = client.run()

    print("[MyCar] End Bot! (PYTHON)")

    exit(return_code)
